File: src/main/java/j/decorator/delegate/Display.java
package j.decorator.delegate;

public final class Display {
  private final DisplayDelegate delegate;

  Display(DisplayDelegate delegate) {
    this.delegate = delegate;
  }

  public int getColumns() {
    return delegate.getColumns();
  }

  public int getRows() {
    return delegate.getRows();
  }

  public String getRowText(int row) {
    return delegate.getRowText(row);
  }

  void show() {
    for (int i = 0; i < getRows(); i++) {
      System.out.println(getRowText(i));
    }
  }
}

File: src/main/scala/s/decorator/delegate/Display.scala
package s.decorator.delegate

object Display {
  def ofString(string: String) = new Display(new StringDisplayDelegate(string))

  def ofSideBorder(display: Display, ch: Char) = new Display(
    new SideBorderDelegate(display, ch)
  )

  def ofFullBorder(display: Display) = new Display(
    new FullBorderDelegate(display)
  )
}

final class Display(private val delegate: DisplayDelegate) {
  def columns: Int = delegate.columns

  def rows: Int = delegate.rows

  def rowText(row: Int): String = delegate.rowText(row)

  private[delegate] def show(): Unit = {
    for (i <- 0 until rows) {
      println(rowText(i))
    }
  }
}
